### PR TITLE
Add Vouched rule [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by marcusoliver**
### Description
Add the Vouched rule to be used a template. 
The rule redirects users to Vouched to complete the identity verification process. When Vouched redirects back via the continue url, the rule saves the result to app metadata. If result already exists in the app metadata, the user is not redirected to Vouched. The rule also determines if the user is allowed to proceed (successfully) through the authentication flow based on the result.

### CODE REVIEW

File: src/rules/vouched-verification.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
+/**
+ * @title Vouched Verification
+ * @overview Verify a person's identity using Vouched.
+ * @gallery true
+ * @category marketplace
+ *
+ * [Vouched](https://vouched.id) 
+ * 
+ * **ID Verification Process**
```
Comment (joshcanhelp) [inline src/rules/vouched-verification.js:9]: TODO: Pull in the description from Marketplace and link to the installation guide.

**Review by seejee — DISMISSED** [review]

**Comment by joshcanhelp** [comment]
@marcusoliver - Thanks for this, it's looking great! Nothing jumps out at me so I'm going to pass this along to our Security group for review.

**Comment by joshcanhelp** [comment]
@marcusoliver - One more thing to note ... our [Rule utility library](https://www.npmjs.com/package/@auth0/rule-utilities) will be available for use tomorrow in production. [You can see an example of how it's used here](https://github.com/auth0/rule-utilities/blob/master/examples/redirectRuleExample.js). There won't be any large changes you'll need to make, just a few to enable some extra checks. When that's ready to include, I'll submit another review calling out the handful of changes to make.

**Comment by marcusoliver** [comment]
Sounds good! Thanks for all your help.

File: src/rules/vouched-verification.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+ */
+
+async function vouchedVerification(user, context, callback) {
+
+    /* ----------- START helpers ----------- */
+    const axios = require('axios');
+    const url = require('url');
+     
+    const defaultApiUrl = 'https://verify.vouched.id/api';
+    const defaultUiUrl = 'https://console.vouched.id';
```
Comment (joshcanhelp) [inline src/rules/vouched-verification.js:24]:
```suggestion
    const defaultUiUrl = 'https://console.vouched.id';
    const idTokenClaim = 'https://vouched.id/is_verified';
```

File: src/rules/vouched-verification.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                }
+  
+                // user is now verified, update app metadata
+                if (isJobVerified(mostRecentJob)) {
+                    user.app_metadata.vouched = extractResults(mostRecentJob);
+                    await auth0.users.updateAppMetadata(user.user_id, user.app_metadata);
+                } else {
+                    // user failed verification check and doesn't have an override
+                    if (configuration.VOUCHED_ID_TOKEN_CLAIM === 'true') {
+                        context.idToken['https://vouched.id/is_verified'] = false;
```
Comment (joshcanhelp) [inline src/rules/vouched-verification.js:149]:
```suggestion
                        context.idToken[idTokenClaim] = false;
```

File: src/rules/vouched-verification.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                url: redirectToVerification(packetId)
+            };
+            return callback(null, user, context);
+        }
+    } catch (e) {
+        return callback(e);
+    }
+  
+    if (configuration.VOUCHED_ID_TOKEN_CLAIM === 'true') {
+        context.idToken['https://vouched.id/is_verified'] = true;
```
Comment (joshcanhelp) [inline src/rules/vouched-verification.js:176]:
```suggestion
        context.idToken[idTokenClaim] = true;
```

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
Alright, the [Rule utility library](https://github.com/auth0/rule-utilities) is ready for use. I left suggestions here, if that helps. After that, I think we're ready to merge this 👍

**Comment by marcusoliver** [comment]
@joshcanhelp i merged your suggestions. that was helpful!
you mentioned previously about adding the integration guide to the header block. is this needed before merging or can it be a subsequent PR?

**Comment by joshcanhelp** [comment]
@marcusoliver - Sorry, I thought I responded here ... we can add it in a later PR 👍  we'll want to standardize how we do that on our end so it will probably be something we can handle for you.

**Review by joshcanhelp — APPROVED** [review]
🎉

File: src/rules/vouched-verification.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    }
+  
+                    return callback(new Error(`This user's ID cannot be verified.`));
+                }
+            }
+        } else {
+            // create Auth0 packet to securely pass info to Vouched
+            const packetId = await createPacket(configuration.VOUCHED_API_KEY, 
+                                                configuration.VOUCHED_PUBLIC_KEY,
+                                                configuration.VOUCHED_CONTINUE_URL,
```
Comment (joshcanhelp) [inline src/rules/vouched-verification.js:169]:
```suggestion
                                                configuration.VOUCHED_CONTINUE_URL,
                                                `https://${context.request.hostname}/continue`
```

File: src/rules/vouched-verification.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                    }
+  
+                    return callback(new Error(`This user's ID cannot be verified.`));
+                }
+            }
+        } else {
+            // create Auth0 packet to securely pass info to Vouched
+            const packetId = await createPacket(configuration.VOUCHED_API_KEY, 
+                                                configuration.VOUCHED_PUBLIC_KEY,
+                                                `https://${context.request.hostname}/continue`
```
Comment (joshcanhelp) [inline src/rules/vouched-verification.js:169]:
```suggestion
                                                `https://${context.request.hostname}/continue`,
```

Comment (joshcanhelp) [inline src/rules/vouched-verification.js:169]:
```suggestion
                                                `https://${context.request.hostname}/continue`,
```

**Review by joshcanhelp — APPROVED** [review]

**Pull request merged**